### ISSUE TITLE
Migrate Battle Notifier store to the Database

### ISSUE THREAD

**Issue by PabloCorso** [elmadev/elmaonline-site#474]
In order to improve the Battle Notifier, as I explained in [this discussion topic](https://github.com/elmadev/elmaonline-site/discussions), it would be great to migrate the current Battle Notifier user configurations in the database from of the current JSON file storage.

This improvement doesn't really add any immediate value to the users but technically it will avoid future confusions because of this quite special way of storage inside this project, and probably make it easier to work with.

I would need help to do this, I'm pretty lost on database action in this project.

The current Battle Notifier saves user configurations for each user in a json file. For example, I set my configurations to listen for any battle by Markku,  plus other random filters:

Discord chat:
```
User: 
!bn
```
```
ElmaOnline (bot): 
[...]
In each line you can match battle types, level names, attributes, duration and designers with the following format:
Battle types, level names ending in .lev (battle attributes between parentheses) >xx <xx by designers
[...]
```
```
User:
First Finish, Apple, M*.lev (see others) >15 <45 by Markku
```

Then the configurations storage is updated:

`bn/bn.store.json`
```json
{
  "219518470767902722": {
    "createdAt": "2020-09-11T20:47:49.039Z",
    "updatedAt": "2021-06-08T19:06:06.969Z",
    "isOn": false,
    "notifyList": [
      {
        "battleTypes": ["First Finish", "Apple"],
        "designers": ["Markku"],
        "levelPatterns": ["M*"],
        "battleAttributes": ["seeOthers"],
        "minDuration": 15,
        "maxDuration": 45
      }
    ],
    "ignoreList": [],
    "username": "pab"
  },
  "other-user-discord-id": { },
  "yet-another-discord-id": { } 
}
```

The `ignoreList` structure is identical to the `notifyList`. To store it we could have the exact same table with a boolean flag to tell if it is a ignore list value.

Can anyone help me get started to where to look for database and how to test things? Not sure how this is working. Also the important bits would be to know how to integrate this with the current Discord login logic.

**Comment by sunehs** [comment]
Discord user id's is in the setting table https://github.com/elmadev/elmaonline-site/blob/dev/src/data/models/Setting.js usage can be seen for example here https://github.com/elmadev/elmaonline-site/blob/6da2ab9327ebe550da1301b6f78be17ebdb82ac8/src/utils/notifications.js#L11

So maybe need to migrate discord id's from bn store to this table. Can add a couple columns to this table if isOn, createdAt etc. is needed, then just need to create one more table for notifyList entries, something like:

KuskiIndex int
BattleTypes string
Designers string
LevelPaterns string
BattleAttributes string
MinDuration int
MaxDuration int
Ignore 1/0

Then look through that table when battle happening to find relevant entries and use KuskiIndex to look up discord id from setting table.

**Comment by sunehs** [comment]
Yea I was just thinking to have them as comma seperated strings. Could also have them as seperate tables, which might be more sql way of doing it, but really depends if there's a benefit. If you're just going to read it like you do now from json store it'll only make things slower and more complicated. Could be useful if set it up so you can get relevant rules with a WHERE instead of looping everything.

**Comment by PabloCorso** [comment]
One issue that I encounter for this migration is that right now Battle Notifier has no idea of `KuskiIndex` or elma.site users for that matter. It just uses a `DiscordId`.

In case we set a `KuskiIndex` for a `BnKuskiRule` entry, then maybe we should force users of BN to sync their user at elma.site with Discord first. Otherwise as today if a user of Discord sets their BN rules via Discord, without sync with elma.site, the server won't know which kuski it is.

The tables might look like this:

```sql
`setting`
KuskiIndex int
BnEnabled 1/0

`bn_kuski_rule`
KuskiIndex int
BattleTypes text
Designers text
LevelPaterns text
BattleAttributes text
MinDuration int
MaxDuration int
Enabled 1/0
Ignored 1/0
```

Setting with a new column to tell if battle notifications is on/off. Removed the info of `createdAt` and `updatedAt` for now.

**Comment by sunehs** [comment]
I think based on the discord user names saved in json store we can guess the KuskiIndex and just update Setting table manually, there's not that many to add.

**Comment by PabloCorso** [comment]
Yes the manual update for the current BN users should be very easy. I mean that the migration will not be as transparent and simple as I thought for the BN discord code, it will have to change a bit to consider the new stored structure. And then new users of BN should sync their Discord first, so that user experience will change too.
